pkg/devicemapper/devmapper_wrapper_dynamic.go

```diff
@@ -2,5 +2,135 @@
 
 package devicemapper
 
-// #cgo pkg-config: devmapper
+/*
+#cgo pkg-config: devmapper
+#cgo LDFLAGS: -ldl
+#include <stdlib.h>
+#include <dlfcn.h>
+#include <libdevmapper.h>
+
+// Yes, I know this looks scary. In order to be able to fill our own internal
+// dm_info with deferred_remove we need to have a struct definition that is
+// correct (regardless of the version of libdm that was used to compile it). To
+// this end, we define struct_backport_dm_info. This code comes from lvm2, and
+// I have verified that the structure has only ever had elements *appended* to
+// it (since 2001).
+//
+// It is also important that this structure be _larger_ than the dm_info that
+// libdevmapper expected. Otherwise libdm might try to write to memory it
+// shouldn't (they don't have a "known size" API).
+struct backport_dm_info {
+	int exists;
+	int suspended;
+	int live_table;
+	int inactive_table;
+	int32_t open_count;
+	uint32_t event_nr;
+	uint32_t major;
+	uint32_t minor;
+	int read_only;
+
+	int32_t target_count;
+
+	int deferred_remove;
+	int internal_suspend;
+
+	// Padding, purely for our own safety.
+	char _[1024];
+};
+
+// We have to wrap this in CGo, because Go really doesn't like function pointers.
+int call_dm_task_deferred_remove(void *fn, struct dm_task *task)
+{
+	int (*_dm_task_deferred_remove)(struct dm_task *task) = fn;
+	return _dm_task_deferred_remove(task);
+}
+*/
 import "C"
+
+import (
+	"unsafe"
+
+	"github.com/sirupsen/logrus"
+)
+
+// dm_task_deferred_remove is not supported by all distributions, due to
+// out-dated versions of devicemapper. However, in the case where the
+// devicemapper library was updated without rebuilding Docker (which can happen
+// in some distributions) then we should attempt to dynamically load the
+// relevant object rather than try to link to it.
+
+// dmTaskDeferredRemoveFct is a "bound" version of dm_task_deferred_remove.
+// It is nil if dm_task_deferred_remove was not found in the libdevmapper that
+// is currently loaded.
+var dmTaskDeferredRemovePtr unsafe.Pointer
+
+// LibraryDeferredRemovalSupport tells if the feature is supported by the
+// current Docker invocation. This value is fixed during init.
+var LibraryDeferredRemovalSupport bool
+
+func init() {
+	// Clear any errors.
+	var err *C.char
+	C.dlerror()
+
+	// We would normally use RTLD_DEFAULT here. But for some reason I've found
+	// that RTLD_DEFAULT doesn't seem to be defined (even with _GNU_SOURCE
+	// enabled) when building cgo. Work around this by opening a handle to the
+	// current process (the semantics are the same).
+	self := C.dlopen(nil, C.RTLD_NOW)
```

Comment:
Yes, or I could just pass `NULL`. I was thinking of doing this, but I'm not sure if `musl` likes that as much as it's not part of POSIX (though I guess since we don't target musl then we don't care).

Reply:
From a quick-n-dirty grep of (musl git)[https://git.musl-libc.org/cgit/musl], it does support it:
```c
include/dlfcn.h:#define RTLD_DEFAULT ((void *)0)
ldso/dynlink.c: if (p == head || p == RTLD_DEFAULT || p == RTLD_NEXT) {
ldso/dynlink.c:         if (p == RTLD_DEFAULT) {
```

So the change I propose would still be valid.

Reply:
Yup, it's done in the newest patchset.